### DISCUSSION TITLE
About note expression

### DISCUSSION THREAD

**Discussion by ammar-mu** [free-audio/clap#328] [Q&A] [answered]
Background:
So I was testing a Seaboard Rise with my soon to be released CLAP plugin. I've noticed that Reaper passes through MIDI expression messages like channel after touch, pitch bend, etc.. from the seaboard almost un-touched to the plugin (CLAP_EVENT_MIDI). While Bitwig (as it seams) translates those messages to CLAP_NOTE_EXPRESSION_ messages. This happens with bitwig even if I use the CLAP_NOTE_DIALECT_MIDI

I don't process CLAP_NOTE_EXPRESSION_ messages. But my plugin can consume the original MIDI events and present them to the user to route as he/she wishes. So probably the only easy/convenient thing for me to do (for bitwig) is to internally translate some of those CLAP_NOTE_EXPRESSION_ back to MIDI. Does that make sense?

Questions:
1. What is CLAP_NOTE_EXPRESSION_EXPRESSION ? 

2. For CLAP_NOTE_EXPRESSION_VOLUME, the spec says the value is 0 < x <= 4. While in some thread here I see it mentioned as 0 to 1. Which one is it?
 
3. I'm probably going to translate CLAP_NOTE_EXPRESSION_VOLUME to MIDI CC7 (Channel Volume). CLAP_NOTE_EXPRESSION_PAN to CC10. CLAP_NOTE_EXPRESSION_TUNING to pitchbend. CLAP_NOTE_EXPRESSION_VIBRATO to CC77. CLAP_NOTE_EXPRESSION_BRIGHTNESS to CC74. And CLAP_NOTE_EXPRESSION_PRESSURE to channel after-touch
Any better choices ?

I could try to figure exactly what CC was the Sea Board sending. But obviously I don't want this translation to be tied to the Sea Board.

Edit: I understand of course that I'll have to internally tie those CC to note_id. So they are per note. I do this using MPE.

**Comment by baconpaul** [comment, marked as answer]
I think bitwig is amiss in doing this translation if you only advertise CLAP_NOTE_DIALECT_MIDI but your choices in #3 seem correct to me.

CLAP_NOTE_EXPRESSION_EXPRESSION is the expression message. You probably want to translate to the BREATH or EXPRESSION PEDAL CC is my guess.

VOLUME is a linear amplitude scaling between 0 and 4, so a VOLUME of 1 matches no change in level, VOLUME of 2 is double amplitude. At least that's what I coded in surge and it callibrated to the display in the bitwig NE viewer OK.

If it helps here's the surge NE handler <https://github.com/surge-synthesizer/surge/blob/main/src/common/dsp/SurgeVoice.cpp> (search for `noteExpression`) and the clap saw demo has it <https://github.com/surge-synthesizer/clap-saw-demo/blob/ea07f4ac4b957d4b53ff4a9a7513a3dabb3a55b6/src/clap-saw-demo.cpp#L635> and <https://github.com/surge-synthesizer/clap-saw-demo/blob/ea07f4ac4b957d4b53ff4a9a7513a3dabb3a55b6/src/saw-voice.cpp#L103>.

**Reply by ammar-mu** [reply]
Thanks a ton for the information and links!!

I'm trying to make Bitwig send me some VOLUME expression to verify that it actually sends values between 0 to 4. (ie. 0 to 1 for no amplification). All I could get is a rare VOLUME set to value=1. No other values were sent,

I mean if the maximum is 4, then the CC for that would translate to 127. Which means 1 would translate to about 32. But I'm still hesitant to do that, thinking that DAWs wouldn't adhere to the 0 to 4 standard. Or they would just not amplify, that is 0 to 1

**Reply by baconpaul** [reply]
You can edit the volume in the note expression editor. Record some MIDI open up the clip and look for the editor then just draw a volume curve 

I don’t recall what MIDI volume spec is but I think making 1 -> 32 seems pretty extreme. Maybe just clamp it out? Surge applies it directly to the vca as a float but if you can’t do that perhaps a more rational thing like 1-> 90 and clamp above 127/90 makes sense?

**Comment by ammar-mu** [comment]
So after I checked that note expression editor. I'm able to receive volume changes. The maximum I get from Bitwig is 2. Thats at 18db gain in the editor. I guess thats just how it is, or may be there is something somewhere I'm missing, I'll look into it.

Yeah, I guess I may need to clamp somewhere, or may be I'll apply some curve to make 0 to 1 take most of the range say from 0 to 100. and 1 to 4 take only the last bits of the range, say 100 to 127.
